stdlib/public/core/Reverse.swift

```diff
@@ -12,34 +12,47 @@
 
 /// An index that traverses the same positions as an underlying index,
 /// with inverted traversal direction.
-public struct ReverseIndex<Base> : Comparable {
-  public init(_ base: Base) { self.base = base }
+public struct ReverseIndex<Base : BidirectionalCollection> : Comparable {
```

Comment:
Why do we need to change the generic parameter type?  This would lead to necessity to box/unbox indices for interop between a collection and its slices.